VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/VitruvioModule.cpp

```diff
@@ -120,29 +120,49 @@ class FLoadResolveMapTask
 	}
 };
 
-void SetInitialShapeGeometry(const InitialShapeBuilderUPtr& InitialShapeBuilder, const TArray<FInitialShapeFace>& InitialShape)
+void SetInitialShapeGeometry(const InitialShapeBuilderUPtr& InitialShapeBuilder, const FInitialShapePolygon& InitialShape)
 {
 	std::vector<double> vertexCoords;
 	std::vector<uint32_t> indices;
 	std::vector<uint32_t> faceCounts;
+	std::vector<uint32_t> holes;
 
-	uint32_t CurrentIndex = 0;
-	for (const FInitialShapeFace& Face : InitialShape)
+	for (const FVector& Vertex : InitialShape.Vertices)
 	{
-		faceCounts.push_back(Face.Vertices.Num());
-		for (const FVector& Vertex : Face.Vertices)
+		const FVector CEVertex = FVector(Vertex.X, Vertex.Z, Vertex.Y) / 100.0;
+		vertexCoords.push_back(CEVertex.X);
+		vertexCoords.push_back(CEVertex.Y);
+		vertexCoords.push_back(CEVertex.Z);
+	}
+
+	for (const FInitialShapeFace& Face : InitialShape.Faces)
+	{
+		faceCounts.push_back(Face.Indices.Num());
+		for (const int32& Index : Face.Indices)
 		{
-			indices.push_back(CurrentIndex++);
+			indices.push_back(Index);
+		}
+
+		if (Face.Holes.Num() > 0)
+		{
+			holes.push_back(faceCounts.size() - 1);
+
+			for (const FInitialShapeHole Hole : Face.Holes)
+			{
+				faceCounts.push_back(Hole.Indices.Num());
+				for (const int32& Index : Hole.Indices)
+				{
+					indices.push_back(Index);
+				}
+				holes.push_back(faceCounts.size() - 1);
+			}
 
-			const FVector CEVertex = FVector(Vertex.X, Vertex.Z, Vertex.Y) / 100.0;
-			vertexCoords.push_back(CEVertex.X);
-			vertexCoords.push_back(CEVertex.Y);
-			vertexCoords.push_back(CEVertex.Z);
+			holes.push_back(std::numeric_limits<uint32_t>::max());
```

Comment:
Is this and line 148 necessary due to how prt handles holes?

Reply:
Yeah exactly. The input must look like this ` The Structure is: { face-with-holes1-idx, hole-face-idx1, hole-face-idx2, ..., MAX_UINT32_T, ..., face-with-holesn-idx, hole-face-idx1, hole-face-idx2, ..., MAX_UINT32_T }`